go-controller/pkg/network-controller-manager/node_network_controller_manager.go

```diff
@@ -70,8 +70,9 @@ func NewNodeNetworkControllerManager(ovnClient *util.OVNClientset, wf factory.No
 	}
 
 	// need to configure OVS interfaces for Pods on secondary networks in the DPU mode
+	// need to start NAD controller on node side for programming gateway pieces for UDNs
 	var err error
-	if config.OVNKubernetesFeature.EnableMultiNetwork && config.OvnKubeNode.Mode == ovntypes.NodeModeDPU {
+	if config.OVNKubernetesFeature.EnableMultiNetwork && config.OvnKubeNode.Mode == ovntypes.NodeModeDPU || util.IsNetworkSegmentationSupportEnabled() {
```

Comment:
hmm no I think because if we did: 
if (config.OVNKubernetesFeature.EnableMultiNetwork || util.IsNetworkSegmentationSupportEnabled()) && config.OvnKubeNode.Mode == ovntypes.NodeModeDPU {}

then we will be invoking a new NAD controller only for DPU mode while we want to always invoke it for primary networks irrespective of whether its dpu or not -> so the whole management port add etc we are doing below is started within this new controller we create, and we want to do that for full mode -> basically only watchdpu stuff should be called for dpu mode rest is only being done for UDNs today

Reply:
(config.OVNKubernetesFeature.EnableMultiNetwork && config.OvnKubeNode.Mode == ovntypes.NodeModeDPU) || util.IsNetworkSegmentationSupportEnabled() is what I want I can make it explicit with the braces

Reply:
but the old code would only add the NAD controller if it was DPU:

if config.OVNKubernetesFeature.EnableMultiNetwork && config.OvnKubeNode.Mode == ovntypes.NodeModeDPU {

why?

Reply:
so IIUC we never did anything related to egress or gateway at all for secondary networks full mode, this is the first time we are doing it. https://github.com/ovn-org/ovn-kubernetes/pull/3398 is where I see the DPU logic was added, the PR has no description and  have no description on https://github.com/ovn-org/ovn-kubernetes/pull/3398/commits/e3ededf1020b3ec3ace1a9d128ec02f6be222015 commit as well. But when I look at `func (bnnc *BaseNodeNetworkController) addRepPort(pod *kapi.Pod, dpuCD *util.DPUConnectionDetails, ifInfo *cni.PodInterfaceInfo, getter cni.PodInfoGetter) error {` it seems we are adding a representator OVS port and use this only in dpu mode?  maybe ? @cathy-zhou or @girishmg can help me here

Reply:
the secondary controller was just used for plugging VF rep into OVS. I now realize @tssurya is extending the controller to do more with UDN.